### PR TITLE
Added provisioning of Default Editor into Devfile

### PR DESCRIPTION
**Pull request by sleshchenko**
### What does this PR do?
The purpose of this PR is provisioning Default Editor into Devfile if there is no another editor. Also default plugins are provided for default editor, even if it was not provisioned automatically but specified by user.

For instance, a user defines the following Devfile:
```yaml
---
specVersion: 0.0.1
name: petclinic-dev-environment
projects:
  - name: petclinic
    source:
      type: git
      location: 'git@github.com:spring-projects/spring-petclinic.git'
```
Then he will get workspace based on the following Devfile
```yaml
---
specVersion: 0.0.1
name: petclinic-dev-environment
projects:
  - name: petclinic
    source:
      type: git
      location: 'git@github.com:spring-projects/spring-petclinic.git'
components:
  - name: theia-editor
    type: cheEditor
    id: org.eclipse.che.editor.theia:1.0.0
  - name: exec-plugin
    type: chePlugin
    id: che-machine-exec-plugin:0.0.1
```
If the user wants to create a workspace without any editor, then he should specify a special attribute for Devfile with name `editorFree`, for example:
```yaml
---
specVersion: 0.0.1
name: petclinic-dev-environment
components:
  - name: myApp
    type: kubernetes
    local: my-app.yaml
attributes:
  - editorFree: true
```

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/12835

#### Release Notes
Added provisioning of Default Editor into Devfile.

#### Docs PR
https://github.com/eclipse/che/pull/13029

### CODE REVIEW

**Comment by sleshchenko** [comment]
My changes are proposed as Draft because there is no final decision about how the user defines a workspace without any editor. Current implementation is 
**1**
```yaml
---
specVersion: 0.0.1
name: petclinic-dev-environment
components:
  - name: no-editor
    type: cheEditor
    id: no-editor
```
As an alternatives ways, it could be implemented as the following
**2**
```yaml
---
specVersion: 0.0.1
name: petclinic-dev-environment
components:
  - name: empty-editor
    type: cheEditor
    id: empty-editor
```
Or 
**3**
```yaml
---
specVersion: 0.0.1
name: petclinic-dev-environment
components:
  ...
attributes:
  editorFree: true
```
@garagatyi @skabashnyuk @mshaposhnik @l0rd Could you please give your feedback about implemented approach for definying workspace without editor, does it look good for you or you prefer alternative ways? Feel free you add more alternative proposals.

**Comment by l0rd** [comment]
@sleshchenko I like number 3 more because it's just one attribute vs 3 lines for an artificial component on the other side

**Comment by skabashnyuk** [comment]
+1 for option 3

File: wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/convert/DefaultEditorProvisioner.java
Comment (metlos) [inline wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/convert/DefaultEditorProvisioner.java:130]: Are we assuming the version can contain colons?

File: wsmaster/che-core-api-devfile/src/main/resources/schema/devfile.json
Comment (metlos) [inline wsmaster/che-core-api-devfile/src/main/resources/schema/devfile.json]: Not sure this is needed. Given the default-editor logic, an empty devfile (i.e. one with no projects and no components) would just be provisioned with the default editor and its supporting plugins. Which I think is a workable environment, isn't it?

**Comment by garagatyi** [comment]
I'm OK with option 3 which seems more popular among voters

**Review by garagatyi — APPROVED** [review]
Looks good

File: wsmaster/che-core-api-devfile/src/main/resources/schema/devfile.json
Comment (sleshchenko) [reply wsmaster/che-core-api-devfile/src/main/resources/schema/devfile.json]: It's kind of strange case to create a workspace with empty Devfile
```
---
specVersion: 0.0.1
name: my-empty-ws
```
But I agree that it makes sense. User will be able to use a workspace with default editor and import projects there. 
Will remove this `anyOf`

File: wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/convert/DefaultEditorProvisioner.java
Comment (sleshchenko) [reply wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/convert/DefaultEditorProvisioner.java:130]: It must not, at least now
https://github.com/eclipse/che/blob/26611a5762af998e09be9c1882a9ba79b7deda96/wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/PluginMetaRetriever.java#L156-L161

I think it makes sense to add field pattern to JsonSchema for such check. @metlos WDYT?

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties
Comment (sleshchenko) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]: Will change to the proposed one. Thanks

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties
Comment (sleshchenko) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties:633]: Not sure if I understand your question.
Here is a check that value is set 
https://github.com/eclipse/che/blob/e0d9742bafa57850c3f0dc9c8672db5a3b1e60c1/wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/convert/DefaultEditorProvisioner.java#L61
Here is converting empty to null
https://github.com/eclipse/che/blob/e0d9742bafa57850c3f0dc9c8672db5a3b1e60c1/wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/convert/DefaultEditorProvisioner.java#L49

Here is converting "NULL" to null
https://github.com/eclipse/che/blob/82b6ea432e57a28f5e07cfc467af563d1ad8cb1b/core/commons/che-core-commons-inject/src/main/java/org/eclipse/che/inject/CheBootstrap.java#L396-L397

Hope it answers your question.

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties
Comment (sleshchenko) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]: Fixed

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties
Comment (metlos) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties:633]: I actually misread the sentence, sorry for the confusion :) I somehow read "that default editor should not be provisioned" as "that default editor should be provisioned".

File: wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/convert/DefaultEditorProvisioner.java
Comment (metlos) [reply wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/convert/DefaultEditorProvisioner.java:130]: That check has not yet been performed at the time we're applying the default editor. Therefore, I think adding the pattern to the schema is a good idea.

**Review by metlos — APPROVED** [review]
LGTM apart from the thing with assuming the version format before we properly checked.

**Comment by sleshchenko** [comment]
ci-test

**Comment by sleshchenko** [comment]
ci-build

**Review by mshaposhnik — APPROVED** [review]

**Comment by sleshchenko** [comment]
ci-test

**Comment by SkorikSergey** [comment]
Selenium tests execution on Eclipse Che Multiuser on OCP (https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/1674//Selenium_20tests_20report/) doesn't show any regression against this Pull Request.

**Pull request merged**
